pkg/kube/kube_handler.go

```diff
@@ -753,24 +778,67 @@ func KubeappsSecretNameForRepo(repoName, namespace string) string {
 	return fmt.Sprintf("%s-%s", namespace, secretNameForRepo(repoName))
 }
 
-func filterAllowedNamespaces(userClientset combinedClientsetInterface, namespaces []corev1.Namespace) ([]corev1.Namespace, error) {
-	allowedNamespaces := []corev1.Namespace{}
-	for _, namespace := range namespaces {
+type checkNSJob struct {
+	ns corev1.Namespace
+}
+
+type checkNSResult struct {
+	checkNSJob
+	allowed bool
+	Error   error
+}
+
+func nsCheckerWorker(userClientset combinedClientsetInterface, nsJobs <-chan checkNSJob, resultChan chan checkNSResult) {
+	for j := range nsJobs {
 		res, err := userClientset.AuthorizationV1().SelfSubjectAccessReviews().Create(context.TODO(), &authorizationapi.SelfSubjectAccessReview{
 			Spec: authorizationapi.SelfSubjectAccessReviewSpec{
 				ResourceAttributes: &authorizationapi.ResourceAttributes{
 					Group:     "",
 					Resource:  "secrets",
 					Verb:      "get",
-					Namespace: namespace.Name,
+					Namespace: j.ns.Name,
 				},
 			},
 		}, metav1.CreateOptions{})
-		if err != nil {
-			return nil, err
+		resultChan <- checkNSResult{j, res.Status.Allowed, err}
+	}
+}
+
+func filterAllowedNamespaces(userClientset combinedClientsetInterface, namespaces []corev1.Namespace) ([]corev1.Namespace, error) {
+	allowedNamespaces := []corev1.Namespace{}
+
+	var wg sync.WaitGroup
+	checkNSJobs := make(chan checkNSJob, maxReq)
+	nsCheckRes := make(chan checkNSResult, maxReq)
+
+	// Process maxReq ns at a time
+	for i := 0; i < maxReq; i++ {
```

Comment:
A couple of questions here:

1. Shouldn't this be only doing `min(maxReq, len(namespaces))` , rather than always `maxReq`? Or,
2. Given that we've already configured the burst of the client to `2*maxReq` perhaps `min(2*maxReq, len(namespaces))`?

That is, I'd expect the number of workers to match the burst, I thought?

Reply:
Yes, I agree, it should be `min(burst,len(namespaces))` however, I'm thinking right now in a multi-user scenario. Since we have configured our k8s client with these burst/qps params, if N users are trying to fetch their namespaces at the same time, we would have a maximum of `N*burst` requests in a single instant.

I have to check, but, do you know what is the behavior if the requests become throttled? Are they queued or an HTTP 429 is thrown? If the latter, we this function perhaps should also implement a kind of backoff mechanism, no?

Reply:
> Yes, I agree, it should be `min(burst,len(namespaces))` however, I'm thinking right now in a multi-user scenario. Since we have configured our k8s client with these burst/qps params, if N users are trying to fetch their namespaces at the same time, we would have a maximum of `N*burst` requests in a single instant.

Yep, theoretically possible. More reason to use the client-go defaults of 10/5

> 
> I have to check, but, do you know what is the behavior if the requests become throttled? Are they queued or an HTTP 429 is thrown? If the latter, we this function perhaps should also implement a kind of backoff mechanism, no?

Haven't dug, but see that there is [support for backoff built-in](https://github.com/kubernetes/client-go/blob/f6ce18ae578c8cca64d14ab9687824d9e1305a67/rest/client.go#L138).

Reply:
I'm using `min(qps, len(ns))` here but note that it doesn't fix any of the other values (the request will take less than a second). This will just limit the number of  connections opened.